Code/MinimalLib/minilib.cpp

```diff
@@ -661,31 +665,40 @@ int JSSubstructLibrary::add_trusted_smiles(const std::string &smi) {
   }
   mol->updatePropertyCache();
   MolOps::fastFindRings(*mol);
-  auto fp = d_fpHolder->makeFingerprint(*mol);
-  if (!fp) {
-    return -1;
+  int smiIdx;
+  if (d_fpHolder) {
+    auto fp = d_fpHolder->makeFingerprint(*mol);
+    if (!fp) {
+      return -1;
+    }
+    int fpIdx = d_fpHolder->addFingerprint(fp);
+    smiIdx = d_molHolder->addSmiles(smi);
+    CHECK_INVARIANT(fpIdx == smiIdx, "");
+  } else {
+    smiIdx = d_molHolder->addSmiles(smi);
   }
-  auto fpIdx = d_fpHolder->addFingerprint(fp);
-  auto smiIdx = d_molHolder->addSmiles(smi);
-  return (fpIdx == smiIdx ? fpIdx : -1);
+  return smiIdx;
 }
 
 int JSSubstructLibrary::add_trusted_smiles_and_pattern_fp(
     const std::string &smi, const std::string &patternFp) {
+  PRECONDITION(d_fpHolder, "d_fpHolder != nullptr");
```

Comment:
Since this function is in the public API, I think this probably should be a `ValueError` exception and not a `PRECONDITION`

Reply:
Done.